Firestore/core/src/firebase/firestore/core/query.cc

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/core/query.h"
+
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
```

Comment:
Does IWYU require that we include types that are part of the interface both in the header and the .cc? Or is it too hard to reason about whether or not the header is required in the .cc if there's any ambiguity?

As it stands I think this file could just `#include ".../query.h"` and nothing else.

Reply:
I actually did a (very) brief survey of both our code base and google3 as I had this same question. It seems like "related header" does not include the interface/parent class, so I've interpreted that IWYU applies.

In a way, this sort of makes sense, since multiple inheritance (or implementing multiple interfaces) implies you could otherwise have multiple related headers... which could themselves also have multiple related headers (and would those count too?) Depending on your interpretation, It gets a little hairy. So I've taken the safe/coward's way out and just re-included everything.

Reply:
I'm fine either way. Note that cpplint includes logic to detect this case, but only handles the first order headers directly included from the class header.

I'd like to be able to articulate a simple rule and then delegate that to tooling so I don't have to think about it in reviews. Maybe I'll just remove that logic from cpplint and call it a day.